File: app/src/main/java/com/ternaryop/widget/CheckableRelativeLayout.java
package com.ternaryop.widget;

import java.util.ArrayList;
import java.util.List;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Checkable;
import android.widget.RelativeLayout;

/**
 * Used to select items in listView
 * 
 * @author dave
 * 
 */
public class CheckableRelativeLayout extends RelativeLayout implements Checkable {
    private boolean isChecked;
    private List<Checkable> checkableViews;

    public CheckableRelativeLayout(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    public CheckableRelativeLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        checkableViews = new ArrayList<Checkable>();
    }

    public boolean isChecked() {
        return isChecked;
    }

    public void setChecked(boolean checked) {
        isChecked = checked;
        setSelected(isChecked);
        for (Checkable c : checkableViews) {
            c.setChecked(checked);
        }
    }

    public void toggle() {
        isChecked = !isChecked;
        setSelected(isChecked);
        for (Checkable c : checkableViews) {
            c.toggle();
        }
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        final int childCount = this.getChildCount();
        for (int i = 0; i < childCount; ++i) {
            findCheckableChildren(this.getChildAt(i));
        }
    }

    private void findCheckableChildren(View v) {
        if (v instanceof Checkable) {
            checkableViews.add((Checkable) v);
        }
        if (v instanceof ViewGroup) {
            final ViewGroup vg = (ViewGroup) v;
            final int childCount = vg.getChildCount();
            for (int i = 0; i < childCount; ++i) {
                findCheckableChildren(vg.getChildAt(i));
            }
        }
    }
}

File: app/src/main/java/com/ternaryop/widget/CheckableRelativeLayout.kt
package com.ternaryop.widget

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.Checkable
import android.widget.RelativeLayout

/**
 * Used to select items in listView
 *
 * @author dave
 */
class CheckableRelativeLayout : RelativeLayout, Checkable {
    private var isChecked: Boolean = false
    private var checkableViews = mutableListOf<Checkable>()

    constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(context, attrs, defStyle)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    override fun isChecked(): Boolean {
        return isChecked
    }

    override fun setChecked(checked: Boolean) {
        isChecked = checked
        isSelected = isChecked
        for (c in checkableViews) {
            c.isChecked = checked
        }
    }

    override fun toggle() {
        isChecked = !isChecked
        isSelected = isChecked
        for (c in checkableViews) {
            c.toggle()
        }
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        val childCount = this.childCount
        for (i in 0 until childCount) {
            findCheckableChildren(this.getChildAt(i))
        }
    }

    private fun findCheckableChildren(v: View) {
        if (v is Checkable) {
            checkableViews.add(v as Checkable)
        }
        if (v is ViewGroup) {
            val vg = v
            val childCount = vg.childCount
            for (i in 0 until childCount) {
                findCheckableChildren(vg.getChildAt(i))
            }
        }
    }
}